### PR TITLE
Feature/add download and metadata

### PR DESCRIPTION
**Pull request by HeikoKarus**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,26 @@
...
+ In order to download the Military MARVEL dataset, download.py a Python script must be run. Set `SAVE_PHOTOGRAPHER = 1` if
+ you also want to save the photographer of each image.
+
+Sort images by class or superclass:
+ In order to sort the images of the Military MARVEL dataset, sort_by.py a Python script must be run. Set `SORT_BY = "class"` if you want to create the fine-grained dataset with 137 classes and `SORT_BY = "superclass"` to create the less fine-grained dataset with 11 classes.
+
+## Cite us
+
+If you use Military MARVEL in your research, please cite:
+
```
Comment (MichaelTeutsch) [inline README.md:32]:
```
@InProceedings{KarusCVPRW2024,   
   Title = {{Towards Explainable Visual Vessel Recognition Using Fine-Grained Classification and Image Retrieval}},   
   Author = {Heiko Karus and Friedhelm Schwenker and Michael Munz and Michael Teutsch},   
   Booktitle = {IEEE CVPR Workshops},   
   Year = {2024}   
}
```

Comment (MichaelTeutsch) [reply README.md:32]: By adding "```" we can make this reference directly copyable.

File: README.md [hunk 2]
```diff
@@ -1 +1,26 @@
-# FineGrainedVesselRecognition
\ No newline at end of file
+# FineGrainedVesselRecognition
+
+This repository shows how to download the Military MARVEL Dataset and sorty it by class or superclass.
```
Comment (MichaelTeutsch) [inline README.md]: "sort it" instead of "sorty it"

File: README.md [hunk 3]
```diff
@@ -1 +1,26 @@
-# FineGrainedVesselRecognition
\ No newline at end of file
+# FineGrainedVesselRecognition
```
Comment (MichaelTeutsch) [inline README.md]: # Towards Explainable Visual Vessel Recognition Using Fine-Grained Classification and Image Retrieval

File: README.md [hunk 4]
```diff
@@ -1 +1,26 @@
...
+
+This repository shows how to download the Military MARVEL Dataset and sorty it by class or superclass.
+
+Here are the direct links to the papers:\
+TODO
+
+The code can be excecuted using Python3.10 and the dependencies can be found in requirments.txt.
```
Comment (MichaelTeutsch) [inline README.md]: "requirements.txt" instead of "requirments.txt"

File: README.md [hunk 5]
```diff
@@ -1 +1,26 @@
...
+
+Downloading Military MARVEL dataset:
+ In order to download the Military MARVEL dataset, download.py a Python script must be run. Set `SAVE_PHOTOGRAPHER = 1` if
```
Comment (MichaelTeutsch) [inline README.md]: I would remove the sentence "Set `SAVE_PHOTOGRAPHER = 1` if  you also want to save the photographer of each image." and rather add a comment line right in the code.

File: README.md [hunk 6]
```diff
@@ -1 +1,26 @@
...
+ you also want to save the photographer of each image.
+
+Sort images by class or superclass:
+ In order to sort the images of the Military MARVEL dataset, sort_by.py a Python script must be run. Set `SORT_BY = "class"` if you want to create the fine-grained dataset with 137 classes and `SORT_BY = "superclass"` to create the less fine-grained dataset with 11 classes.
+
```
Comment (MichaelTeutsch) [inline README.md:28]: Maybe an example for a ground truth label or the file structure?

File: sort_by.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                os.makedirs(category_dir)
+
+            shutil.copy(
+                os.path.join(IMAGE_DIR, image_id + ".jpg"),
+                os.path.join(category_dir, image_id + ".jpg"),
+            )
+
+
+def main():
+
```
Comment (MichaelTeutsch) [inline sort_by.py]: I would remove this empty line.

File: download.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            ) as text_file:
+                text_file.write(image_data["page_data"]["print_photographer"])
+
+    except HTTPError:
+        # Image was deleted
+        pass
+
+
+def main():
+
```
Comment (MichaelTeutsch) [inline download.py]: I would remove this empty line.

File: README.md [hunk 9]
```diff
@@ -1 +1,26 @@
-# FineGrainedVesselRecognition
\ No newline at end of file
+# FineGrainedVesselRecognition
+
+This repository shows how to download the Military MARVEL Dataset and sorty it by class or superclass.
+
+Here are the direct links to the papers:\
+TODO
+
```
Comment (MichaelTeutsch) [inline README.md:6]: Maybe add here the way to install the dataset:
1. create Conda env
2. pip install reqs
3. run download.py

File: README.md [hunk 10]
```diff
@@ -1 +1,26 @@
-# FineGrainedVesselRecognition
\ No newline at end of file
+# FineGrainedVesselRecognition
+
+This repository shows how to download the Military MARVEL Dataset and sorty it by class or superclass.
+
+Here are the direct links to the papers:\
```
Comment (MichaelTeutsch) [inline README.md]: Here is the direct link to the paper:

File: download.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+from multiprocessing import Pool as ThreadPool
+from urllib.request import Request, urlopen
+from urllib.error import HTTPError
+from pathlib import Path
+import re
+
+from bs4 import BeautifulSoup
+import tqdm
+
+SAVE_PHOTOGRAPHER = 0
```
Comment (MichaelTeutsch) [inline download.py]: More documentation. Shift the sentence about SAVE_PHOTOGRAPHER from the readme.md to here. Please document NUMBER_OF_WORKERS. What is "FILE_TO_DOWNLOAD_FROM"?

File: download.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+FILE_TO_DOWNLOAD_FROM = "metadata.json"
+SOURCE_LINK = "https://www.shipspotting.com/photos/"
+IMAGE_DIR = os.path.join(os.getcwd(), "data", "images")
+META_DIR = os.path.join(os.getcwd(), "data", "metadata")
+
+
+def download_image(file_id: str):
+    """This function downloads an image from a given URL and saves it
+    to a specified directory. It also saves the author's name in a text
+    file if specified. If the image is deleted, it will not save anything.
```
Comment (MichaelTeutsch) [inline download.py]: Suggestion: "If the image was deleted on the original website, this image will be skipped and nothing will be downloaded or saved."

File: download.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    try:
+        url = SOURCE_LINK + file_id
+        req = Request(url, headers={"User-Agent": "Mozilla/5.0"})
+        html = urlopen(req, timeout=300).read()
+        soup = BeautifulSoup(html, "lxml")
+
+        images = [img for img in soup.findAll("img")]
+        image_links = [each.get("src") for each in images]
+
+        for each in image_links:
```
Comment (MichaelTeutsch) [inline download.py]: Maybe rather "image_link" instead of "each" as variable name?

File: download.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            image_data = json.loads(
+                data_script.text.split("window._INITIAL_DATA =")[1].strip()[:-1]
+            )
+            with open(
+                os.path.join(META_DIR, file_id + ".txt"), "w", encoding="utf-8"
+            ) as text_file:
+                text_file.write(image_data["page_data"]["print_photographer"])
+
+    except HTTPError:
+        # Image was deleted
```
Comment (MichaelTeutsch) [inline download.py:84]: Would it make sense to write a comment to the log that image xy was not found and probably was deleted on the original website?

File: download.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+    # Make dir for images files
+    if not os.path.exists(META_DIR) and SAVE_PHOTOGRAPHER:
+        os.makedirs(META_DIR)
+
+    # Only download files that are not there
+    downloaded_ids = [file.stem for file in Path(IMAGE_DIR).glob("*.jpg")]
+    image_ids = list(set(image_ids) - set(downloaded_ids))
+
+    # Make the Pool of workers
```
Comment (MichaelTeutsch) [inline download.py]: pool with lower case p

File: sort_by.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
+import os
+import sys
+from pathlib import Path
+import json
+import shutil
+
+
+SORT_BY = "class"  # or "superclass"
```
Comment (MichaelTeutsch) [inline sort_by.py]: Some more documentation of the constants here.

File: sort_by.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+IMAGE_DIR = os.path.join(os.getcwd(), "data", "images")
+FILE_TO_DOWNLOAD_FROM = "metadata.json"
+
+
+def create_sorted_dataset(
+    downloaded_ids: list[str],
+    annotation_data: list[dict[str, str]],
+    dataset_folder: str,
+):
+    """The create_sorted_dataset function loops through each image_id in downloaded_ids and searches for the
```
Comment (MichaelTeutsch) [inline sort_by.py:37]: This documentation of the function should be in line 12, or not?

Comment (HeikoKarus) [reply sort_by.py:37]: Position of docstring is correct

**Pull request merged**